compiler/src/dotty/tools/dotc/core/TypeUtils.scala

```diff
@@ -128,22 +128,19 @@ class TypeUtils:
         case None => throw new AssertionError("not a tuple")
 
     def namedTupleElementTypesUpTo(bound: Int, derived: Boolean, normalize: Boolean = true)(using Context): List[(TermName, Type)] =
+      def extractNamesTypes(nmes: Type, vals: Type): List[(TermName, Type)] =
+        val names = nmes.tupleElementTypesUpTo(bound, normalize).getOrElse(Nil).map(_.dealias).map:
+          case ConstantType(Constant(str: String)) => str.toTermName
+          case t => throw TypeError(em"Malformed NamedTuple: names must be string types, but $t was found.")
+        val values = vals.tupleElementTypesUpTo(bound, normalize).getOrElse(Nil)
+        names.zip(values)
+        
       (if normalize then self.normalized else self).dealias match
         // for desugaring and printer, ignore derived types to avoid infinite recursion in NamedTuple.unapply
-        case AppliedType(tycon, nmes :: vals :: Nil) if !derived && tycon.typeSymbol == defn.NamedTupleTypeRef.symbol =>
-          val names = nmes.tupleElementTypesUpTo(bound, normalize).getOrElse(Nil).map(_.dealias).map:
-            case ConstantType(Constant(str: String)) => str.toTermName
-            case t => throw TypeError(em"Malformed NamedTuple: names must be string types, but $t was found.")
-          val values = vals.tupleElementTypesUpTo(bound, normalize).getOrElse(Nil)
-          names.zip(values)
+        case defn.NamedTupleDirect(nmes, vals) => extractNamesTypes(nmes, vals)
```

Comment:
```suggestion
        case defn.NamedTupleDirect(nmes, vals) if !derived => extractNamesTypes(nmes, vals)
```

right?  Or is the removal intentional?

Reply:
I think it's equivalent, since if `NamedTupleDirect` matches then we're done, the value of`derived` doesn't matter. With your change it will just be checked in a slightly different order, since `NamedTuple.unapply` calls `NamedTupleDirect.unapply` as its base case.